chore: fill README (#29)
strong-config/node: README.md

# 💪 strong-config

<!-- markdownlint-disable line-length -->
[![pipeline status](https://git.brickblock.sh/devops/strong-config-ts-mirror/badges/master/pipeline.svg)](https://git.brickblock.sh/devops/strong-config-ts-mirror/commits/master) [![coverage report](https://git.brickblock.sh/devops/strong-config-ts-mirror/badges/master/coverage.svg)](https://git.brickblock.sh/devops/strong-config-ts-mirror/commits/master)
<!-- markdownlint-enable line-length -->

> Secret management for NodeJS. Based on [SOPS](https://github.com/mozilla/sops).

Reviewer: I know this was here before, but I do think we should put **config** first and **secrets** second. Because it has utility even if you weren't using it for managing secrets, plus it's called strong-config :)

```suggestion
Simple and secure config management for NodeJS. Based on [SOPS](https://github.com/mozilla/sops).
```

Author: Happy to put `simple`. Not sure about the `secure` - it's misleading imo as it reads as the management is secure.

`Simple config-with-secrets management for NodeJS` ?

Reviewer: my argument would be that it _is_ more secure in a few ways:
- It's harder to accidentally check in secrets
- It's using SOPS which makes using strong cryptography much easier

Author: Hm yeah. Your first point is more safety, not security.

Still feeling weird about `secure management` as a first-sentence selling point. I'll check the SOPS repo and see how they market themselves to get some inspiration

Have you ever pushed secrets to your git remote by accidentially committing

> the `.env` file? Then `strong-config` (built by Brickblock) is here to help:

Reviewer: Good start, I would add some more examples here maybe to make this point stronger.

You should find some content in my [blog post drafts](https://www.notion.so/brickblock/Part-1-env-files-suck-3f5e70e599a04f889790fc49dd831caa)

Author: Ouh first time hearing about those drafts. Thx

- Manage configs for multiple runtime environments, e.g. different clusters
- Encrypt secrets at rest and automatically decrypt them at runtime
- Validate configs against a custom schema
- For Typescript users: automatically generate types based on schema

## How-to

1. Install `strong-config` in your project

    ```sh
    npm install @strong-config/node
    yarn add @strong-config/node
    ```

1. Define runtime environment and create first config file

    `strong-config` uses an environment variable to determine which config file
    to load. By default, this is `RUNTIME_ENVIRONMENT`.

>     ```sh

Reviewer: i'm generally a fan of code-block first, _then_ explanation or description. think about yourself, usually you just wanna get going and get the code, right?

Author: generally agree but this is a one-liner. no strong opinion.

I think it's good style to first introduce concepts before you use them. The same way you explain an acronym the first time you use it. So not having a problem with one-liner explanation and then code block.

>     export RUNTIME_ENVIRONMENT=development # or similar, depending on your shell

Reviewer: I don't quite understand why we need this line here, what am I missing?

Author: Will soon be `NODE_ENV` instead but why wouldn't we need it? Gotta tell strong-config which config to load.

I wouldn't set `NODE_ENV=development` implicitly so the user isn't aware. Shouldn't mess with the user's env vars

Reviewer: hmm, but why would this need to be set in the context of creating a new folder and a new file?

`mkdir config` and `echo 'blub' > config/development.yaml` both don't access `NODE_ENV` in any way or am I missing something?

Author: Are are you referring to setting `NODE_ENV` in a separate point?

I'm not opposed to the current draft because it shows the mapping between file name and `NODE_ENV`. In this light, both steps (setting NODE_ENV, and, creating a dir+file) fit together

Reviewer: Yeah, my main point is that to me it feels the wrong place to set the env var because it isn't needed for this. To me as an engineer i would ask myself: "Why would I need to set an env var for creating a folder?"

IMO we should set it once we actually use it

Author: Since you have a very clear how this README should look like, where should that step go? It must be fore the JS block.

Just gonna ask to save us both some time

    # in project root

>     mkdir config

Reviewer: ```suggestion
# Create a ./config folder in the root of your project
mkdir config
```

Author: IMO too much. Someone worrying about config+secrets management in NodeJS _does_ know what `mkdir` does.

My style is compact here. The less text the better in the first `How-to`.

    echo "someField: '💪'" > config/development.yaml
    ```

    By setting `RUNTIME_ENVIRONMENT=development`, the config file
  `development.yaml` will be loaded from the config directory.

1. Import and use `strong-config` in code

>     ```js

Reviewer: I like to have a comment at the top to clarify which file we're talking about.

```suggestion
    ```js
    // ./src/index.js (or in whichever file you'd like to load your configuration)
```

Author: Also here i'd go for smth more _from engineer to engineer_ type of language. How about:

```
// whereever you want to use config values
```
or
```
// app.js, or wherever you want to use config values
```

Reviewer: Fine with your suggestion 👍

    // ES6-style import. Commonjs is supported too
    import StrongConfig from '@strong-config/node'

    // uses default parameters
    const config = (new StrongConfig()).load()

    console.log(config)
    // --> { someField: '💪' }
    ```

    The first invocation of `load()` dynamically generates types for Typescript and
    saves them to `./strong-config.d.ts`. If you do not use Typescript or despise
    types because of other reasons, feel free to delete it and disable type
    generation completely (see [Customization](#customization)).

## Customization

`strong-config` can be customized to your needs. Out of the box, sensible
defaults are used. However, these can be easily overwritten by passing an
`options`-object to the class contructor:

```js
const strongConfig = new StrongConfig(<options>)
```

The available options and their defaults are:

```js
{
  TODO
}
```

If you do not want to generate types, set `types: false`.

## Schema Validation

> yadayada, schema.json, path defaults, how to overwrite, link json-schema,

Reviewer: perfect 😀

      created_at: '2019-11-07T10:11:12Z'
      enc: AiAAmdAgjOdw1XdV2MsDpvmZ4Deo867hmcX2B3DNhe2BCF2axuZ18hJJFK9oBlE1BrD70djwqi+L8T+NRNVnGUP+1//w8cJATAfJ8W/cQZFcdFTqjezC+VYv9xYI8i1bRna4xfFo/INIJtFDR38ZH1nrQg==
  lastmodified: '2019-11-07T10:11:12Z'
  mac: ENC[AES256_GCM,data:ABcd1EF2gh3IJKl4MNOpQr5stuvWXYz6sBCDEfGhIjK=,iv:A1AaAAAaa111a1Aa111AA/aaaAaaAAaa+aAaAaAAAaA=,tag:AAaaA1a1aaaAa/aa11AaaA==,type:str]
  encrypted_suffix: Secret
  version: 3.4.0
```

The field `sops` contains a set of metadata provided context information and
important details to decrypt this config at runtime. Never delete this field
or make changes to the encrypted file without using `@strong-config` - it will
leave you with a corrupted config.

## How-to

Please backup your existing configs and secrets.

1. Install `@strong-config/node` and the SOPS binary.

   ```sh
   npm install @strong-config/node
   # or
   yarn add @strong-config/node
   ```

   SOPS is [available for all major systems](https://github.com/mozilla/sops/releases).
   If you are on Mac, you can install SOPS by running `brew install sops`.

1. Define runtime environment and create first config file

   ```sh

>    export NODE_ENV=development # or similar, depending on your shell

Reviewer: To finish the discussion off from the previous thread: Still think this is confusing here.

Because you're assuming that the shell they're currently in will also be the one that they're going to call `yarn start` from. Which is not guaranteed.

I think what's missing is a last step in the how-to that's putting the NODE_ENV variable and the starting of the application into context:

```sh
# Important: NODE_DEV determines which file in the ./config folder will be loaded at runtime
NODE_DEV=development yarn start 
```

Author: Hm we have more assumptions

- people use the same shell or a sub-shell for all steps
- people use `yarn start` as a single entry point to the app
- people use bash or zsh

I just added a last step to the README. Let me know what you think. I believe we can't make everyone happy.

Author: Thoughts, @crosby ?

Reviewer: Yeah, you can never make everyone happy. But javascript developers being familiar with `yarn start` is one of the safest assumptions to make in my book.

We should also not give examples for other shells than bash IMO, for the same reason. Bash is the default and everyone using zsh/fish/powershell etc will be familiar with translating bash commands to their respective shell.

if you remove the fish example from the new last step then: LGTM 👍

Author: Then we should change everything `yarn` to `npm` too right? https://www.npmtrends.com/yarn-vs-npm

Author: removed the fish line

Reviewer: IMO in the end it's ok that `strong-config` also reflects _our opinions_ a little (which is currently preferring `yarn` over `npm`).

Plus, stats are tricky: https://trends.google.de/trends/explore?q=yarn,npm

to validate your configs. The schema file must be written in JSON according to the
[`json-schema`](https://json-schema.org/) standard. To get started, you can have
a look at the [official learning resources](https://json-schema.org/learn/) or
checkout the examples in the `/example` directory of this project.

You can provide the path to your schema file via `options.schemaPath`. Please
note that this file must be named `schema.json` if you decide to place it in
the same directory as your config files (which is the default).

However, `strong-config` will work fine if you decide to not use schemas at all.

## FAQ

1. **Can I write my configs as JSON files?**

   Yes, YAML and JSON are supported. Please pay attention to the file extensions
   (`.yaml|yml` or `.json` are allowed).

1. **Do you support schemas other than [`json-schema`](https://json-schema.org/)?**

   No, currently we only support a schema based on `json-schema`.

1. **Can I use `@strong-config/node` for the browser, too?**

   No, single page applications (SPAs) are not suitable for containing secrets.
   This is because SPA code is delivered to the client and is therefore always public.

>    TODO: explain what you can do when your configs don't have secrets.

Reviewer: overlooked a TODO here

Author: expanded this one

Reviewer: 👍

1. **Can I use `@strong-config/node` for Server-Side-Rendered (SSR) apps, too?**

   It depends. The server-side can make use of `@strong-config/node` to manage
   secrets. However, the client-side can't because of the reason in the previous
   point.

1. **Can my config contain arrays?**

   Yes, but not as top-level definition. The top-level of any config file must be
   an object so SOPS can store its metadata in the field `sops`. Otherwise, you
   can define arrays at any nested level of the config file.

## Development and Contributions

We added a few scripts (under `/scripts/dev` with entrypoints in `package.json`)
for development purposes:

- `yarn dev:load`/`yarn dev:load:es6`, and `yarn dev:load:commonjs`
- `yarn dev:validate`

These use an exemplary config setup in `/example`.

> Our contribution guidelines are posted in `CONTRIBUTING.md`

Reviewer: This won't be true until we actually have a `CONTRIBUTING.md`, I'd delete it until we actually have it, otherwise it's just a dead link

Author: right. deleted

    ```sh
    npm install @strong-config/node
    yarn add @strong-config/node
    ```

1. Define runtime environment and create first config file

    ```sh
    export NODE_ENV=development # or similar, depending on your shell

    # inside your project
    mkdir config
    echo "someField: '💪'" > config/development.yaml
    ```

    `strong-config` uses an environment variable to determine which config file
    to load. By default, this is `NODE_ENV`. Setting `NODE_ENV=development` will
    eventually load `development.yaml` from the config directory.

1. Import and use `strong-config` in code

    ```js
    // index.js, or anywhere you need config values

    // ES6-style import. Commonjs is supported too
    import StrongConfig from '@strong-config/node'

    // uses default parameters
    const config = (new StrongConfig()).load()

    console.log(config)
    // --> { someField: '💪' }
    ```

    The first invocation of `load()` dynamically generates types for Typescript and
    saves them to `./strong-config.d.ts`. If you do not use Typescript, feel
    free to delete it and disable type generation completely (see [Customization](#customization)).

1. ```sh

>    yarn remove dotenv

Reviewer: This seems a bit random here, it's the only time we ever mention `dotenv` in this doc so it's weird introducing it here

Author: is more a joke. i removed it

Reviewer: in this case: gotta work on your jokes a little more 🤡

# 💪 strong-config

<!-- markdownlint-disable line-length -->
[![pipeline status](https://git.brickblock.sh/devops/strong-config-ts-mirror/badges/master/pipeline.svg)](https://git.brickblock.sh/devops/strong-config-ts-mirror/commits/master) [![coverage report](https://git.brickblock.sh/devops/strong-config-ts-mirror/badges/master/coverage.svg)](https://git.brickblock.sh/devops/strong-config-ts-mirror/commits/master)
<!-- markdownlint-enable line-length -->

Simple, safe and secure config management for NodeJS. Based on [SOPS](https://github.com/mozilla/sops).

Have you ever struggled with different config values across local, CI,

> and various cluster environments? If no, did you ever forget to update the

Reviewer 2: if ~~no~~ not

Author: done

> and various cluster environments? If no, did you ever forget to update the

Reviewer 2: We shouldn't presume they're using clusters here. Just "various environments" is fine.

Author: done

cluster config after updating the local config? Have you ever been worried about

> the secrecy of your DB credentials and API keys taking the `.env` file approach?

Reviewer 2: "by taking the"

Author: done

> Ever leaked confidential data by pushing secrets in your `.env` file to the git remote?

Reviewer 2: "Have you ever"

Author: done

> Do you agree that just `dotenv` is not enough? **`strong-config` is here to help!**

Reviewer 2: Instead of a rhetorical question here it should IMO be "If any of this has happened to you, **`strong-config` is here to help!**"

Author: done

> - Structure configs hierarchically by writing JSON or YAML.

Reviewer 2: This list should be prefixed with something e.g. "Strong Config allows you to:"

Author: updated

> - Manage different configs for different environments, e.g. different clusters

Reviewer 2: Again, remove the "clusters" here.

Reviewer: Agree with the general comment of removing "cluster" references. I'd say it's safe to assume that most JS devs are _not_ heavy Kubernetes users so this isn't the best example

Author: I'm sure a JS devs knows what a cluster is without having heard of Kubernetes.

Agreed though, removed all `cluster`s

- Protect your secrets by applying strong cryptography. Encrypt secrets at rest
  and automatically decrypt them at runtime by integrating a key management
  service (KMS), e.g. [AWS KMS](https://aws.amazon.com/kms/) or
  [Google Cloud KMS](https://cloud.google.com/kms/).

> - Enforce arbitrary permission models provided by your KMS. Decide who can encrypt

Reviewer 2: Arbitrary is the wrong word here I think. It sounds like you mean that your assigning random permissions.

Author: not sure if `any` is better. I removed it now.

>   and decrypt configs for which cluster environments.

Reviewer 2: Remove "cluster" here.

Reviewer 2: "Decide who can encrypt and decrypt secrets per environment"

> - Optionally validate them against a custom schema using

Reviewer 2: ~~them~~ configuration

Author: updated

>   [`json-schema`](https://json-schema.org/). Also in CI.

Reviewer 2: `json-schema` doesn't validate, it just defines the schema. Perhaps it's better to say "Optionally validate configuration against a custom schema defined using [`json-schema`](https://json-schema.org/)"

Author: updated

> - Safeguard your configs by using git hooks. Make sure the configs are both

Reviewer 2: Not sure this is a feature of this library.

Author: hmm, the list is prefixed with 'Strong Config allows you to:', so I'd say this point can be there

  valid and encrypted before committing and pushing them.

> - Autogenerate types based on your schema. These types can be imported in your

Reviewer 2: Might be best to say "typescript types" here.

>   sources to write safer code (Typescript only for now).

Reviewer 2: Not sure it needs the clarification about importing into source code for now.

Author: updated

> These benefits help you to catch inconsistencies, exposed secrets, and invalid

Reviewer 2: Not sure about this whole paragraph. The benefits are already clear in the list above, a summary paragraph feels a little superfluous.

Author: deleted

configs as early as possible in the development process. To do so, you will use
widely adopted tools you are probably already familiar with: JSON (or YAML),
your cloud provider, and yet another npm package. Try it out!

> 

Author: So, I elaborated on this part. IMO a bit too much text now early on. When I see a README, I want to have text that is easier to digest. On the other hand, I don't want to make this a science

## Example

Here is how an encrypted config file containing secrets looks like:

```yml
logger:
  level: DEBUG
auth:
  url: my-auth-url
  myApiKeySecret: ENC[AES256_GCM,data:aeQ+hlVIah7WyJoVR/Jbkb6GLH7ihsV0D81+U++pkiWD0zeoRL/Oe9Q3Tz6j/TNvKKVDnohIMyw3UVjELOuSY+A==,iv:nVRZWogV4B7o=,tag:KrE2jssfP4uCvqq+pc/JyQ==,type:str]
sops:
  gcp_kms:
  -   resource_id: projects/my-project/locations/europe-west3/keyRings/my-project-key-ring/cryptoKeys/my-strong-config-key
      created_at: '2019-11-07T10:11:12Z'
      enc: AiAAmdAgjOdw1XdV2MsDpvmZ4Deo867hmcX2B3DNhe2BCF2axuZ18hJJFK9oBlE1BrD70djwqi+L8T+NRNVnGUP+1//w8cJATAfJ8W/cQZFcdFTqjezC+VYv9xYI8i1bRna4xfFo/INIJtFDR38ZH1nrQg==
  lastmodified: '2019-11-07T10:11:12Z'
  mac: ENC[AES256_GCM,data:ABcd1EF2gh3IJKl4MNOpQr5stuvWXYz6sBCDEfGhIjK=,iv:A1AaAAAaa111a1Aa111AA/aaaAaaAAaa+aAaAaAAAaA=,tag:AAaaA1a1aaaAa/aa11AaaA==,type:str]
  encrypted_suffix: Secret
  version: 3.4.0
```

The field `sops` contains a set of metadata provided context information and
important details to decrypt this config at runtime. Never delete this field

> or make changes to the encrypted file without using `@strong-config` - it will

Reviewer 2: Jokes are fine and it **is** a problem to delete this value, but we don't want to scare people off too much here.

Author: yeah we dont want to scare people. done

leave you with a corrupted config and a world full of terror.

## How-to

Please backup your existing configs and secrets_

1. Install `@strong-config/node` in your project

    ```sh
    npm install @strong-config/node
    yarn add @strong-config/node
    ```

1. Define runtime environment and create first config file

    ```sh
    export NODE_ENV=development # or similar, depending on your shell

    # inside your project
    mkdir config
    echo "someField: '💪'" > config/development.yaml
    ```

    `strong-config` uses an environment variable to determine which config file
    to load. By default, this is `NODE_ENV`. Setting `NODE_ENV=development` will
    eventually load `development.yaml` from the config directory.

1. Import and use `strong-config` in code

    ```js
    // index.js, or anywhere you need config values

    // ES6-style import. Commonjs is supported too
    import StrongConfig from '@strong-config/node'

    // uses default parameters
    const config = (new StrongConfig()).load()

    console.log(config)
    // --> { someField: '💪' }
    ```

    The first invocation of `load()` dynamically generates types for Typescript and

>     saves them to `./strong-config.d.ts`. If you do not use Typescript or despise

Reviewer 2: I'd remove the "or despise types because of other reasons"

Author: done
